sphinx/ext/intersphinx.py

```diff
@@ -533,17 +534,90 @@ def run(self) -> tuple[list[Node], list[system_message]]:
         assert self.name == self.orig_name.lower()
         inventory, name_suffix = self.get_inventory_and_name_suffix(self.orig_name)
         if inventory and not inventory_exists(self.env, inventory):
-            logger.warning(__('inventory for external cross-reference not found: %s'),
-                           inventory, location=(self.env.docname, self.lineno))
+            self._emit_warning(
+                __('inventory for external cross-reference not found: %s'), inventory
+            )
             return [], []
 
-        role_name = self.get_role_name(name_suffix)
+        domain_name, role_name = self._get_domain_role(name_suffix)
+
         if role_name is None:
-            logger.warning(__('role for external cross-reference not found: %s'), name_suffix,
-                           location=(self.env.docname, self.lineno))
+            self._emit_warning(
+                __('invalid external cross-reference suffix: %s'), name_suffix
+            )
             return [], []
 
-        result, messages = self.invoke_role(role_name)
+        # attempt to find a matching role function
+        role_func: None | RoleFunction
```

Comment:
```suggestion
        role_func: RoleFunction | None
```

I observed that you like putting None before the other types. Is it a Rust reason?

Reply:
Not much reason to be honest, would prefer ruff to just sort it for me lol,
maybe I feel it follows more in the "old" `Optional[xxx]` (and for rust `Option<xxx>`) syntax, whereby you are saying up-front that something is "optional"

Reply:
why do you want to put `None` last?

Reply:
First reason is that it's what we have mostly in typeshed iirc and what's being told on Python docs (https://docs.python.org/3/library/typing.html#typing.Optional).

Second reason is that you can see "differences" between types faster, namely `T | None` vs `U | None`, you see that you have T and U as types, whereas `None | T` and `None | U`, you first see "None" as a "common" possibility.